modules/extensions-support/src/main/java/org/mule/module/extension/internal/manager/DefaultExtensionManager.java

```diff
@@ -0,0 +1,425 @@
+                                                          operationContext.getOperation().getName(), extension.getName(), matches));
+        }
+        else
+        {
+            attemptToCreateImplicitConfigurationInstance(extension, operationContext);
+            return getOperationExecutor(operationContext);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OperationExecutor getOperationExecutor(final String configurationInstanceProviderName, OperationContext operationContext)
+    {
+        ConfigurationInstanceProvider<Object> configurationInstanceProvider = getConfigurationInstanceProvider(configurationInstanceProviderName);
+        return getOperationExecutor(configurationInstanceProvider, operationContext);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<Extension> getExtensions()
+    {
+        return register.getExtensions();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public <C> Set<Extension> getExtensionsCapableOf(Class<C> capabilityType)
+    {
+        checkArgument(capabilityType != null, "capability type cannot be null");
+        return register.getExtensionsCapableOf(capabilityType);
+    }
+
+    private <C> void registerConfigurationInstance(Configuration configuration, String configurationInstanceName, C configurationInstance)
+    {
+        ExtensionStateTracker extensionStateTracker = register.getExtensionState(configuration);
+        extensionStateTracker.registerConfigurationInstance(configuration, configurationInstanceName, configurationInstance);
+
+        putInRegistryAndApplyLifecycle(configurationInstanceName, configurationInstance);
+    }
+
+    private void putInRegistryAndApplyLifecycle(String key, Object object)
+    {
+        try
+        {
+            muleContext.getRegistry().registerObject(objectNameHelper.getUniqueName(key), object);
+        }
+        catch (MuleException e)
+        {
+            throw new MuleRuntimeException(e);
+        }
+    }
+
+    private OperationExecutor getOperationExecutor(ConfigurationInstanceProvider<Object> configurationInstanceProvider, OperationContext operationContext)
+    {
+        Object configurationInstance = getConfigurationInstance(configurationInstanceProvider, operationContext);
+        OperationExecutor executor;
+
+        synchronized (configurationInstance)
+        {
+            ExtensionStateTracker extensionStateTracker = register.getExtensionState(configurationInstanceProvider.getConfiguration());
+            executor = extensionStateTracker.getOperationExecutor(configurationInstanceProvider.getConfiguration(), configurationInstance, operationContext);
+            if (executor == null)
+            {
+                executor = createOperationExecutor(configurationInstance, operationContext);
```

Comment:
The synchronized block should be here, adding another check for executor == null inside